### ISSUE TITLE
redefinition error in type checkers

### ISSUE THREAD

**Issue by cmukesh19** [wesselb/plum#46]
```python
from plum import dispatch

@dispatch
def add_nums(num1: int, num2: int) -> int:
    return num1 + num2

@dispatch
def add_nums(num1: str, num2: str) -> int:
    return add_nums(int(num1), int(num2))

add_nums(10, 15)
add_nums('30', '25')
```

flake8 reports `redefinition of unused add_nums from line 4 flake8(F811)`. The code works fine though.  

Another issue (I was using VS Code), editor always shows the last defined function in the hint. If there are multiple overloads. ![VS Code hints](https://i.imgur.com/IccCsP0.jpg)

**Comment by wesselb** [comment]
Unfortunately, `flake8` isn't aware of this package and doesn't support the multiple dispatch pattern. Here's a [relevant thread](https://github.com/PyCQA/pyflakes/issues/320). At this point, I think you'll have to add `# noqa: F811` comments to every function definition to make `flake8` stop complaining. I agree that this is suboptimal. Ideally, we'd have `flake8` support the multiple dispatch pattern.

As for the suggestions of VS Code, I'm unfortunately not aware of how precisely that is working. It is possible that this could be fixed by maintaining all implementations of a function in a way that can be accessed by intellisense.

**Comment by cmukesh19** [comment]
Makes sense. Python's new typing library and this library is amazing to get nice structure to the code. I dislike the idea of one function taking various parameters, and then applying intermediate code within its body. I am already fan of plum 👍 

As you suggested, while flake8, implements something, I'll ignore check F811. VS Code allows to add it centrally, so no need to put it in each function. Not ideal, but nice tradeoff.

For function signature, I think this could be similar issue as the first one. I'll raise it in VS Code forum. For now, I have used `@overload` from typing library and that solved it. Little verbose, but still nice. Full implementation will look like:

```python
from plum import dispatch
from typing import overload

@overload
def add_nums(num1: int, num2: int) -> int:
    ...

@overload
def add_nums(num1: str, num2: str) -> int:
    ...

@dispatch
def add_nums(num1: int, num2: int) -> int:
    return num1 + num2

@dispatch
def add_nums(num1: str, num2: str) -> int:
    return add_nums(int(num1), int(num2))

add_nums(10, 15)
```
`overload` definitions should always be declared before `@dispatch`.  VS Code will show ![VS Type hint](https://i.imgur.com/hUwrJuF.jpg)

Note: not done much testing

Thank you @wesselb

**Comment by PhilipVinc** [comment]
In theory we could have plum automatically declare the `@overload` when a function is decorated?

**Comment by PabloLION** [comment]
@PhilipVinc 
I tried to add static types for other packages (called selenium) before. VS Code buried it deep tho. It's hidden in 
https://github.com/python/typeshed/tree/master/stubs of https://github.com/python/typeshed in
https://github.com/microsoft/pyright/tree/main/packages/pyright-internal of https://github.com/microsoft/pyright in
https://github.com/microsoft/pylance-release

I'd be more than appreciated if you can work on it. Thanks in advance!❤️

**Comment by PhilipVinc** [comment]
@PabloLION I'm sorry, I don't understand what you are talking about and what you are proposing?

**Comment by PabloLION** [comment]
My bad. I didn't made it clear..
I was using VSCode too but with PyLance and it reports a similar of `Method declaration "XXX" is obscured by a declaration of the same name`. The error was thrown because in first link above there's no folder called "plum" so pylance doesn't know that `plum.dispatch` is actually taking care of the dispatch / overload. I assume flake8 also uses the same source to lint the code.
So my suggestion was to type the `plum` in that repo `python/typeshed` and then the linters knows `plum.dispatch` works like the built-in `@typing.overload`.
Sorry again for the inconvenience.
